src/firmware/boards/legacy_robot_stm32f4/io/receive.c

```diff
@@ -191,107 +200,79 @@ void receive_tick(log_record_t *record)
     }
 }
 
-
-/**
- * \brief Returns the serial number of the most recent drive packet.
- */
-uint8_t receive_last_serial(void)
-{
-    return __atomic_load_n(&last_serial, __ATOMIC_RELAXED);
-}
-
-void handle_drive_packet(uint8_t *dma_buffer)
+void handle_drive_packet(uint8_t *packet_data, size_t packet_size)
 {
-    // Grab emergency stop status from the end of the frame.
-    bool estop_run =
-        !!dma_buffer[MESSAGE_PAYLOAD_ADDR + NUM_ROBOTS * RECEIVE_DRIVE_BYTES_PER_ROBOT];
-
-    // Grab a pointer to the robot’s own data block.
-    const uint8_t *robot_data =
-        dma_buffer + MESSAGE_PAYLOAD_ADDR + RECEIVE_DRIVE_BYTES_PER_ROBOT * robot_index;
+    uint8_t packet_robot_id                 = packet_data[packet_size - 3];
+    bool estop_triggered                    = packet_data[packet_size - 2] != 1;
+    bool feedback_requested_from_this_robot = packet_data[packet_size - 1] == robot_index;
 
     // Check if feedback should be sent.
-    if (robot_data[0] & 0x80)
+    if (feedback_requested_from_this_robot)
     {
         feedback_pend_normal();
     }
 
-    // Extract the serial number.
-    uint8_t serial = *robot_data++ & 0x0F;
-
-    // Construct the individual 16-bit words sent from the host.
-    uint16_t words[4U];
-    for (unsigned int i = 0U; i < 4U; ++i)
+    // Check if this drive packet was intended for us, if not we can stop here
+    if (packet_robot_id != robot_index)
     {
-        words[i] = *robot_data++;
-        words[i] |= (uint16_t)*robot_data++ << 8;
+        return;
     }
 
-    // In case of emergency stop, treat everything as zero
-    // except the chicker discharge bit (that can keep its
-    // status).
-    if (!estop_run)
+    // Reset timeout.
+    timeout_ticks = 1000U / portTICK_PERIOD_MS;
+
+    // Figure out what primitive to run
+    primitive_params_t pparams = {0};
+    unsigned int primitive     = 0;
+    if (estop_triggered)
+    {
+        // Set the primitive to be a stop primitive
```

Comment:
Is there another way we can implement this since it's a *safety* feature?  If the stop primitive values are changed for whatever reason this will stop working. Maybe some kind of function or defines that are linked to the stop primitive?

Reply:
So again, we definitely should have a better solution here, but for now this isn't a regression, so in the interests of limiting the scope of this PR I'm going to say we should leave it as-is for now.